File: src/twenty/november/no102/BinaryTreeLevelOrderTraversal.java
package twenty.november.no102;

import java.util.*;

public class BinaryTreeLevelOrderTraversal {
    public static void main(String[] args) {
        TreeNode root1 = new TreeNode(3, new TreeNode(9), new TreeNode(20, new TreeNode(5), new TreeNode(7)));
        TreeNode root2 = new TreeNode(1, new TreeNode(2, new TreeNode(3, new TreeNode(4, new TreeNode(5, new TreeNode(6), null), null), null), null), null);
        TreeNode root3 = null;
//        List<List<Integer>> result = levelOrderTemplate(root3);
        List<List<Integer>> result = levelOrder2(root2);
        result.forEach(System.out::println);
    }

    private static List<List<Integer>> levelOrder2(TreeNode root) {
//        use dfs
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        dfs(root, result, 0);
        return result;
    }

    private static void dfs(TreeNode root, List<List<Integer>> result, int level) {
        if (result.size() < level + 1) {
            result.add(new ArrayList<>());
        }
        result.get(level).add(root.val);
        if (root.left != null) {
            dfs(root.left, result, level + 1);
        }
        if (root.right != null) {
            dfs(root.right, result, level + 1);
        }
    }

    private static List<List<Integer>> levelOrderTemplate(TreeNode root) {
        if (root == null) {
            return Collections.emptyList();
        }
        List<List<Integer>> result = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            int currentLevel = queue.size();
            List<Integer> levelResult = new ArrayList<>();
            for (int i = 0; i < currentLevel; i++) {
                TreeNode node = queue.poll();
                levelResult.add(node.val);
                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
            }
            result.add(levelResult);
        }
        return result;
    }

    private static List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null) {
            return Collections.emptyList();
        }
        Deque<TreeNode> queue = new LinkedList<>();
        queue.offerLast(root);
        TreeNode last = root;
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> levelResult = new ArrayList<>();
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            levelResult.add(node.val);
            if (node.left != null) {
                queue.offerLast(node.left);
            }
            if (node.right != null) {
                queue.offerLast(node.right);
            }
            if (node == last) {
                result.add(levelResult);
                levelResult = new ArrayList<>();
                last = node.right != null ? node.right : node.left != null ? node.left : queue.peekLast();
            }

        }
        return result;
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

/*
* 给你一个二叉树，请你返回其按 层序遍历 得到的节点值。 （即逐层地，从左到右访问所有节点）。

 

示例：
二叉树：[3,9,20,null,null,15,7],

    3
   / \
  9  20
    /  \
   15   7
返回其层次遍历结果：

[
  [3],
  [9,20],
  [15,7]
]

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/binary-tree-level-order-traversal
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
* */

File: go/year2025/month3/no102/BinaryTreeLevelOrderTraversal.go
package main

import "fmt"

func main() {
	root1 := &TreeNode{Val: 3,
		Left: &TreeNode{Val: 9},
		Right: &TreeNode{Val: 20,
			Left:  &TreeNode{Val: 15},
			Right: &TreeNode{Val: 7}},
	}
	root2 := &TreeNode{Val: 1}
	root3 := (*TreeNode)(nil)
	println(fmt.Sprintf("%+v", levelOrder(root1))) // [[3],[9,20],[15,7]]
	println(fmt.Sprintf("%+v", levelOrder(root2))) // [[1]]
	println(fmt.Sprintf("%+v", levelOrder(root3))) // []
}

// BFS, Time: O(n), Space: O(n)
func levelOrder(root *TreeNode) [][]int {
	ans := make([][]int, 0)
	if root == nil {
		return ans
	}
	queue := make([]*TreeNode, 0)
	queue = append(queue, root)
	for len(queue) > 0 {
		level := make([]int, 0)
		nextQueue := make([]*TreeNode, 0)
		for _, node := range queue {
			level = append(level, node.Val)
			if node.Left != nil {
				nextQueue = append(nextQueue, node.Left)
			}
			if node.Right != nil {
				nextQueue = append(nextQueue, node.Right)
			}
		}
		ans = append(ans, level)
		queue = nextQueue
	}
	return ans
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/*
 给你二叉树的根节点 root ，返回其节点值的 层序遍历 。 （即逐层地，从左到右访问所有节点）。

示例 1：
输入：root = [3,9,20,null,null,15,7]
输出：[[3],[9,20],[15,7]]

示例 2：
输入：root = [1]
输出：[[1]]

示例 3：
输入：root = []
输出：[]

提示：
树中节点数目在范围 [0, 2000] 内
-1000 <= Node.val <= 1000
*/
